proposals/ytt/003-openapi-schemas/README.md

```diff
@@ -0,0 +1,80 @@
+
+# Problem Statement
+Configuration Authors want to be able to generate documentation for configuration inputs (for example, Data Values Schema, and Data Values) for their users. Additionally, Configuration Consumers want to be able to validate their configuration inputs by other tools (for example, IDE's and OpenAPI Schema validators). 
+
+# Proposal
+Implement a flag to export an [OpenAPI Schema](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md) yaml document generated from a [Data Values schema](https://carvel.dev/ytt/docs/latest/lang-ref-ytt-schema/) file, and an optional Data Values file. When invoking ytt with configuration inputs and the flag `--schema-inspect-openapi`, the result is the OpenAPI formatted document only. 
+
+This flag will behave similar to the `--data-values-inspect` flag, which excludes any templates in the output.
+
+## Examples:
+### Scenario 1: Data Values Schema file only
+`ytt -f schema.yml --schema-inspect-openapi` will generate an OpenAPI Schema document using the values from the schema.yml file.
+
+### Scenario 2: Data Values Schema and Data Values
+`ytt -f schema.yml -f values.yml --schema-inspect-openapi` will generate an OpenAPI Schema document using the merged values of both the schema.yml and the values.yml file.
+
+Note: Using the merged result of both files allows the generated OpenAPI Schema to have a default value for array items, since Data Values Schema files do not yet support default values for arrays. Additionally, it allows the OpenAPI Schema to document the merged result of Data Values Schema and Data Values that will ultimately be used as configuration input during templating.
+
+### Scenario 3: Data Values file only
+`ytt -f values.yml --schema-inspect-openapi` will result in a warning suggesting that a Data Values Schema file be provided. 
+
+The reason for waiting to support Data Values files only is so that the OpenAPI document can have the additional information of default values and descriptions only available in Data Values Schema files. If this functionality comes with no additional work, we should reevaluate this decision.
+
+# Specification
+## Glossary
+* OpenAPI Schema: A document that follows this [spec](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#openapi-object)
+* Collection: a yaml map or array
+* Configuration inputs: a ytt Schema file and a Data Values file
+* [Data Value Schema](https://carvel.dev/ytt/docs/latest/lang-ref-ytt-schema/) A ytt document that declares default values and type information for parameterized values.
+* [Data Value](https://carvel.dev/ytt/docs/latest/ytt-data-values/) A ytt document that overrides defaults declared in a Data Values Schema.
+
+
+## Generated Openapi Document Spec:
+
+### Root Document fields:
+```yaml
+openapi: 3.1.0 # Version of the openapi spec
+info:
+  version: 1.0.0 # Version of this document
+  title: Openapi schema generated from ytt schema # Title of this document
```

Comment:
would we want the title to include a variable portion indicating e.g. filename(s)/paths/ ytt-command-line that was used to assemble?   Can the document provide enough breadcrumbs for a reader / naive maintainer to know how it was generated and how to regenerate or update it?

Reply:
Yes, this is a good idea. 
At bare minimum a user could name the file in a way that correlates it to the inputs to ytt (`schema.openapi.yml`?). 

But it does become less straightforward to just add the name of the Data Values Schema file to the title because we may use an arbitrary amount of Data Values Schema and Data Values file all merged together to come up with the result..

Reply:
should/could we add a metadata section that has even something as crude as `generedated-by: ${args}` ? - if we dump all the command line args into a file is there ever a risk that we'd expose a secret or a credential that was passed in as a data value?      I'm still looking for guidance re. making it easy to reproduce / regenerate / maintain,  vs. the tension of complications.

Reply:
> generedated

stealing this verb. 😜

Reply:
> metadata section that has even something as crude as `generedated-by: ${args}` ?

even better, the OpenAPI specification itself allows for ad-hoc extensions, requiring such fields to be prefixed with `x-` ([ref](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#specificationExtensions)).

So, `x-generedated-by: ${args}`

Reply:
> Can the document provide enough breadcrumbs for a reader / naive maintainer to know how it was generated and how to regenerate or update it?

I believe that in most cases, the invocation of `ytt` to generate the schema will be straight-forward:

```console
$ ytt -f config/ --data-values-schema-inspect-openapi
```

Is that true?


On the implementation side, when we're generating the output, things like command-line args and source files are not in hand... so mating up those sources with the output would introduce some complexity.

Reply:
The command to generate a schema should be mostly straightforward, just `ytt -f config/ --data-values-schema-inspect` since it does not accept any additional flags right now. And @pivotaljohn is correct, we lose the information of what command was run.

Since an OpenAPI schema is generated from Data Values Schema at a specific point in time (or at a specific sha), it should be included in version control with the rest of the configuration, and running the command with the same config will yield the same OpenAPI Document. I wonder if there is a way to tie the version of the OpenAPI document to a version of the config it was generated from as a way to make it more clear how it was created..